MobileWallet/Backup/ICloudBackup.swift

```diff
@@ -0,0 +1,342 @@
+    func didFinishUploadBackup(percent: Double, completed: Bool, error: Error?)
+}
+
+class ICloudBackup: NSObject {
+
+    var query: NSMetadataQuery!
+
+    private let containerIdentifier = "iCloud.com.tari.wallet"
+    private let directory = TariLib.shared.databaseDirectory
+    private let fileName = "Tari-Aurora-Backup"
+    private var observers = NSPointerArray.weakObjects()
+
+    var inProgress: Bool = false
+    private(set) var progressValue: Double = 0.0
+
+    static let shared = ICloudBackup()
+
+    override init() {
+        super.init()
+        initialiseQuery()
+        addNotificationObservers()
+    }
+
+    func initialiseQuery() {
+        query = NSMetadataQuery.init()
+        query.operationQueue = .main
+        query.searchScopes = [NSMetadataQueryUbiquitousDataScope]
+        query.predicate = NSPredicate(format: "%K LIKE '*.zip'", NSMetadataItemFSNameKey)
+    }
+
+    func addObserver(_ observer: ICloudBackupObserver) {
+        observers.addObject(observer)
+    }
+
+    // returns true if backup of current wallet is exist
+    func isBackupExist() -> Bool {
+        let fileManager = FileManager.default
+        guard let backupFolder = TariLib.shared.tariWallet?.publicKey.0?.hex.0 else { return false }
+
+        do {
+            let icloudFolderURL = try iCloudDirectory()
+
+            if let urls = try? fileManager.contentsOfDirectory(at: icloudFolderURL, includingPropertiesForKeys: nil, options: []) {
+                if let _ = urls.first(where: { $0.absoluteString.contains(backupFolder) }) { return true }
+            }
+        } catch {
+            return false
+        }
+
+        return false
+    }
+
+    func createWalletBackup() throws {
+        guard let backupFolder = TariLib.shared.tariWallet?.publicKey.0?.hex.0 else { throw ICloudBackupWalletError.iCloudContainerNotFound }
+
+        let fileURL = try zipBackupFiles()
+        let containerURL = try iCloudDirectory().appendingPathComponent(backupFolder)
+
+        if !FileManager.default.fileExists(atPath: containerURL.path) {
+            try FileManager.default.createDirectory(at: containerURL, withIntermediateDirectories: true, attributes: nil)
+        }
+        let backupFileURL = containerURL.appendingPathComponent(fileURL.lastPathComponent)
+        if FileManager.default.fileExists(atPath: backupFileURL.path) {
+            try FileManager.default.removeItem(at: backupFileURL)
+            try FileManager.default.copyItem(at: fileURL, to: backupFileURL)
+        } else {
+            try FileManager.default.copyItem(at: fileURL, to: backupFileURL)
+        }
+        query.operationQueue?.addOperation({ [weak self] in
+            _ = self?.query.start()
+            self?.query.enableUpdates()
+        })
+    }
+
+    func restoreWallet(completion: () -> Void) throws {
+        Keychain.logout()
+        let wallets = try existsWallets()
+        guard let firstWallet = wallets.first else { return }
```

Comment:
You can combine these two lines into a single function that could read better with a name like `getLastICloudWalletBackup`, if I understood correctly what it's doing.

Or if you will keep it in two lines, a better name could help with the function, such as `getICloudWalletBackups`.